File: quicktap-stub/src/main/java/com/comrade/quicktaprelay/QuickTapRelayActivity.java
package com.comrade.quicktaprelay;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

public final class QuickTapRelayActivity extends Activity {
    private static final String CYBERSYN_PACKAGE = "com.termux.cybersyn";
    private static final String ACTION_EXTERNAL_TRIGGER = "com.termux.cybersyn.action.EXTERNAL_TRIGGER";
    private static final String EXTRA_TRIGGER_NAME = "com.termux.cybersyn.extra.TRIGGER_NAME";
    private static final String EXTRA_SOURCE = "com.termux.cybersyn.extra.SOURCE";
    private static final String EXTRA_TRIGGER_TIME_MS = "com.termux.cybersyn.extra.TRIGGER_TIME_MS";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Intent relay = new Intent(ACTION_EXTERNAL_TRIGGER)
                .setPackage(CYBERSYN_PACKAGE)
                .putExtra(EXTRA_TRIGGER_NAME, "quick_tap")
                .putExtra(EXTRA_SOURCE, "pixel_quick_tap_stub")
                .putExtra(EXTRA_TRIGGER_TIME_MS, System.currentTimeMillis());
        sendBroadcast(relay);
        finish();
        overridePendingTransition(0, 0);
    }
}

File: app/src/main/java/com/termux/cybersyn/core/external/QuickTapRelayActivity.kt
package com.termux.cybersyn.core.external

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.core.content.ContextCompat
import com.termux.cybersyn.core.contexts.ExternalTriggerContextEvents
import com.termux.cybersyn.core.engine.AutomationService
import com.termux.cybersyn.core.logging.AppLogger

class QuickTapRelayActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        relayQuickTap()
        finish()
        overridePendingTransition(0, 0)
    }

    private fun relayQuickTap() {
        val now = System.currentTimeMillis()
        sendBroadcast(
            Intent(ExternalTriggerContract.ACTION_EXTERNAL_TRIGGER).apply {
                setPackage(packageName)
                putExtra(ExternalTriggerContract.EXTRA_TRIGGER_NAME, ExternalTriggerContextEvents.DEFAULT_TRIGGER)
                putExtra(ExternalTriggerContract.EXTRA_SOURCE, ExternalTriggerContract.SOURCE_PIXEL_QUICK_TAP)
                putExtra(ExternalTriggerContract.EXTRA_TRIGGER_TIME_MS, now)
            },
        )
        runCatching {
            ContextCompat.startForegroundService(
                this,
                Intent(this, AutomationService::class.java)
                    .putExtra(AutomationService.EXTRA_STARTED_FROM_VISIBLE_UI, false),
            )
        }.onFailure { error ->
            AppLogger.error(TAG, "Failed to start automation service after Quick Tap", error)
        }
        AppLogger.debug(TAG, "Relayed Pixel Quick Tap")
    }

    companion object {
        private const val TAG = "QuickTapRelay"
    }
}
